File: src/Urql.Player.Compat/Session/PlayerFrame.cs
using Urql.Core.Diagnostics;
using Urql.Core.Runtime;
using Urql.Player.Compat.RichText;
using Urql.Player.Compat.Skin;
using Urql.Player.Compat.Viewport;

namespace Urql.Player.Compat.Session;

public sealed record PlayerFrame(
    int VirtualWidth,
    int VirtualHeight,
    ViewTransform ViewTransform,
    IReadOnlyList<RichRun> TextRuns,
    IReadOnlyList<FrameButton> Buttons,
    IReadOnlyList<FrameMenu> Menus,
    IReadOnlyList<Diagnostic> Diagnostics,
    VmStatus Status,
    bool HitInstructionLimit,
    SkinDefinition Skin);

public sealed record FrameButton(int Id, string Caption, string Target);

public sealed record FrameMenu(string Name, IReadOnlyList<string> Entries);

File: web/packages/player/src/player-frame.ts
import type { Diagnostic } from "@fireurq/core";
import type { VmStatus } from "@fireurq/core";

export interface PlayerTextRun {
  kind: "text";
  text: string;
}

export interface FrameButton {
  id: number;
  caption: string;
  target: string;
}

export interface ViewTransform {
  virtualWidth: number;
  virtualHeight: number;
  viewWidth: number;
  viewHeight: number;
  scale: number;
  offsetX: number;
  offsetY: number;
}

export interface PlayerFrame {
  virtualWidth: number;
  virtualHeight: number;
  viewTransform: ViewTransform;
  textRuns: PlayerTextRun[];
  buttons: FrameButton[];
  diagnostics: Diagnostic[];
  status: VmStatus;
  hitInstructionLimit: boolean;
}
